File: assets/codes/leetcode/src/main/java/leetcode/L_98.go
package sample

var stackNode []int

func isValidBST(root *TreeNode) bool {
	stackNode = []int{}
	helper(root)
	if len(stackNode) == 1 {
		return true
	}
	for i := 1; i < len(stackNode); i++ {
		if stackNode[i] <= stackNode[i-1] {
			return false
		}
	}
	return true
}

func helper(root *TreeNode) {
	if root == nil {
		return
	}
	helper(root.Left)
	stackNode = append(stackNode, root.Val)
	helper(root.Right)
}

File: assets/codes/leetcode/src/main/java/leetcode/L_98.java
package leetcode;

public class L_98 {
    long pre = Long.MIN_VALUE;
    boolean b = true;

    public boolean isValidBST(TreeNode root) {
        dfs(root);
        return b;
    }

    public void dfs(TreeNode root) {
        if (root == null) {
            return;
        }
        dfs(root.left);
        if (root.val > pre) {
            pre = root.val;
        } else {
            b = false;
            return;
        }
        dfs(root.right);
    }

    public static void main(String[] args) {
        L_98 l98 = new L_98();
        {
            TreeNode treeNode = new TreeNode(-2147483648);
            System.out.println(l98.isValidBST(treeNode));
        }
        {
            TreeNode treeNode = new TreeNode(2);
            treeNode.left = new TreeNode(1);

            treeNode.right = new TreeNode(3);
            System.out.println(l98.isValidBST(treeNode));
        }
        {
            TreeNode treeNode = new TreeNode(5);
            treeNode.left = new TreeNode(1);

            treeNode.right = new TreeNode(4);
            treeNode.right.left = new TreeNode(3);
            treeNode.right.right = new TreeNode(6);
            System.out.println(l98.isValidBST(treeNode));
        }
    }
}
